### PR TITLE
Paginated transactions

### PR DESCRIPTION
**Pull request by bodymindarts**
Builds off of https://github.com/GaloyMoney/galoy/pull/447 to support paginated transactions in the new schema

### CODE REVIEW

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (nicolasburtey) [inline src/app/wallets/get-transactions-for-wallet.ts:37]: may need to add an additional error to let the client know that not all the data has been returned

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (nicolasburtey) [inline src/app/wallets/get-transactions-for-wallet.ts:46]: `confsLT` not clear to me what LT means. not less than I guess?

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (nicolasburtey) [inline src/app/wallets/get-transactions-for-wallet.ts:54]: I think we may never want `price = 0` should be a fallback value. for multiple reasons:
- if we need to do a conversion, we may need to divide by this value. this would lead to further issues
- it could be some form of an attack vector eventually for USD account

Comment (nicolasburtey) [reply src/app/wallets/get-transactions-for-wallet.ts:54]: so I think at this point we either want to have `NaN` or trigger an error.

Comment (nicolasburtey) [reply src/app/wallets/get-transactions-for-wallet.ts:54]: maybe that doesn't matter here because it's only for pending transaction

File: src/domain/bitcoin/onchain/tx-decoder.ts
Comment (nicolasburtey) [inline src/domain/bitcoin/onchain/tx-decoder.ts:25]: may not be needed

File: src/domain/wallets/tx-history.ts
Comment (nicolasburtey) [inline src/domain/wallets/tx-history.ts:15]: may want to use lodash for this one

File: src/domain/wallets/tx-history.ts
Comment (nicolasburtey) [inline src/domain/wallets/tx-history.ts:86]: may need to differentiate between intraledger initiated by username and initiated by lightning invoice

File: src/graphql/types/object/fiat-wallet.ts
Comment (nicolasburtey) [inline src/graphql/types/object/fiat-wallet.ts:33]: debug code?

File: src/domain/wallets/tx-history.ts
Comment (bodymindarts) [reply src/domain/wallets/tx-history.ts:86]: We've been discussing this a lot. Lets finalise our approach with @jrbuhl93 when he arrives

**Comment by dolcalmi** [comment]
please change the base, if we don't merge #447 check this is a duplicate effort

**Comment by bodymindarts** [comment]
> please change the base, if we don't merge #447 check this is a duplicate effort

Yeah we are directly working on this branch... will close the other one

File: src/domain/ledger/index.types.d.ts
Comment (vindard) [inline src/domain/ledger/index.types.d.ts:13]: I believe we can type this directly here to get the constant values assigned to each field. Typescript should allow something like `type a = "Alice"` to be type string restricted to single string value `'Alice'`.

```suggestion
type LedgerTransactionType = {
  Invoice: "invoice"
  Payment: "payment"
  IntraLedger: "on_us"
  LnFeeReimbursement: "fee_reimbursement" // lightning
  OnchainReceipt: "onchain_receipt"
  OnchainPayment: "onchain_payment"
  OnchainIntraLedger: "onchain_on_us"
  OnchainDepositFee: "deposit_fee" // onchain
  Fee: "fee"
  Escrow: "escrow"
  RoutingFee: "routing_fee" // channel-related
  ExchangeRebalance: "exchange_rebalance" // send/receive btc from the exchange
  UserRebalance: "user_rebalance" // buy/sell btc in the user wallet
  ToColdStorage: "to_cold_storage"
  ToHotWallet: "to_hot_wallet"
}
```

File: src/services/mongoose/schema.types.d.ts
Comment (vindard) [inline src/services/mongoose/schema.types.d.ts:78]: This is also defined in `domain/accounts/index.types.d.ts`
```suggestion
```

File: src/domain/accounts/index.types.d.ts
Comment (vindard) [inline src/domain/accounts/index.types.d.ts:2]: I believe we can type this directly here to get the constant values assigned to each field, although I'm not sure if we deliberately did it this way so that the type is defined in only one place.

If not, typescript should allow something like `type a = "Alice"` to be type string restricted to single string value `'Alice'`.

_(This comment would be for all places where we use this pattern)_

```suggestion
type AccountLevelType = {
  One: 1,
  Two: 2,
}
```

File: src/domain/accounts/index.ts
Comment (vindard) [inline src/domain/accounts/index.ts:4]: If we implement the type in the `.d.ts` file, this can probably change. Also, would this be a good candidate for a typescript `enum`?

```suggestion
export const AccountLevel: AccountLevelType = {
  One: 1,
  Two: 2,
}
```

File: src/domain/wallets/tx-history.ts
Comment (vindard) [inline src/domain/wallets/tx-history.ts:105]: Probably doesn't need an `else` here because of early return in previous block and consistency with Ln 86.

```suggestion
      }
      if (addresses && addresses.length > 0) {
        return {
```

File: test/integration/02-user-wallet/02-send-lightning.spec.ts
Comment (vindard) [inline test/integration/02-user-wallet/02-send-lightning.spec.ts:379]: In places like this, do we also want to check for something like `expect(user2OnUsTxn[0].old.type).toBe("on_us")` since we're doing it in other places?

File: test/integration/02-user-wallet/02-send-lightning.spec.ts
Comment (vindard) [inline test/integration/02-user-wallet/02-send-lightning.spec.ts:131]: A comment I added that can probably be removed now that it looks like the test works with the new changes.
```suggestion
```

**Review by vindard — CHANGES_REQUESTED** [review]
I added some small cleanup suggested changes and some questions around typing constants.

File: src/domain/accounts/index.types.d.ts
Comment (bodymindarts) [reply src/domain/accounts/index.types.d.ts:2]: What I want is to be able to reference the constants in code without typing the strings:
```
const level = AccountLevelType.One
```

if we put the definition in the `.types.d.ts` file we would have the definition in 2 places or have to hard code values everywhere:
```
const level = 1
```

enums have other drawbacks as they are not native to JS https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums.

All things considered I chose this pattern for collections of constants but perhaps we can discuss it in the team at daily.

File: test/integration/02-user-wallet/02-send-lightning.spec.ts
Comment (bodymindarts) [reply test/integration/02-user-wallet/02-send-lightning.spec.ts:379]: Should probably only check for the consistently `settlementVia` but don't want to clean it all up now.

File: src/domain/wallets/tx-history.ts
Comment (vindard) [inline src/domain/wallets/tx-history.ts:91]: There was some logic on the `main` branch that suggested that `Intraledger` payments could have `paymentHash` values as well and that this was used as the unique identifier for the `invoiceusers` table (see [this test line](https://github.com/GaloyMoney/galoy/blob/548ac0fc673280033fef8f7bff3bbb68e7401baf/test/integration/02-user-wallet/02-send-lightning.spec.ts#L68) for e.g.).

If we changed that behavior then this is ok, but if not then this could potentially give us `Intraledger` transactions as well as `Lightning` transactions I think.

**Review by vindard — COMMENTED** [review]
~All looks good, just one question before I approve~

File: src/domain/wallets/tx-history.ts
Comment (vindard) [reply src/domain/wallets/tx-history.ts:91]: Oh nevermind, I just caught the rest of logic just under this... LGTM :+1:

**Review by vindard — APPROVED** [review]

File: src/domain/wallets/tx-history.ts
Comment (bodymindarts) [reply src/domain/wallets/tx-history.ts:91]: yes that is why there is a switch on the `settlementVia` field:
```
      if (paymentHash) {
        return {
          id,
          initiationVia: PaymentInitiationMethod.Lightning,
          settlementVia:
            type === LedgerTransactionType.IntraLedger
              ? SettlementMethod.IntraLedger
              : SettlementMethod.Lightning,
```

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (dolcalmi) [inline src/app/wallets/get-transactions-for-wallet.ts:20]: why this method is using a different pattern for return?

**Review by dolcalmi — DISMISSED** [review]

File: src/domain/users/index.types.d.ts
Comment (bodymindarts) [reply src/domain/users/index.types.d.ts:12]: yes that's right

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (bodymindarts) [reply src/app/wallets/get-transactions-for-wallet.ts:20]: Because it can return partial results. ie getting ledger transactions works then we want to pass those back to the user but also want to return the error to communicate that we couldn't get the pending txs.

File: src/graphql/types/object/user.ts
Comment (bodymindarts) [reply src/graphql/types/object/user.ts:21]: I don't understand your question.

A user has a default account (one of many accounts). Account has this type:
```
type Account = {
  id: AccountId
  level: AccountLevel
  status: AccountStatus
  walletIds: WalletId[]
}
```

so to get to your wallet you go through the query:
```
me -> defaultAccount -> wallet
```

Comment (bodymindarts) [reply src/graphql/types/object/user.ts:21]: All the values for `User`, `Account` and `Wallet` are currently on the same `Users` table of mongoose - but data migration is out of scope of this PR.

File: src/graphql/types/object/ln-transaction.ts
Comment (bodymindarts) [reply src/graphql/types/object/ln-transaction.ts]: done

File: src/graphql/types/object/onchain-transaction.ts
Comment (bodymindarts) [reply src/graphql/types/object/onchain-transaction.ts]: done

File: src/graphql/types/object/user.ts
Comment (dolcalmi) [reply src/graphql/types/object/user.ts:21]: thanks

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (dolcalmi) [reply src/app/wallets/get-transactions-for-wallet.ts:20]: uhmm, so if my inner code can return a partial result then should I use this "pattern"? 

you use a lot of wrappers in other parts of the code, why is this better  than use a wrapper and have consistency in the way that we handle errors?

Comment (bodymindarts) [reply src/app/wallets/get-transactions-for-wallet.ts:20]: > you use a lot of wrappers in other parts of the code, why is this better than use a wrapper and have consistency in the way that we handle errors?

can you show an example here?

Comment (bodymindarts) [reply src/app/wallets/get-transactions-for-wallet.ts:20]: Just discussing this with @samerbuna.
Think we could standardise on returning an array for partial results:
```
[ResultType, ErrorType]
```

Comment (dolcalmi) [reply src/app/wallets/get-transactions-for-wallet.ts:20]:
```js
export const getTransactionsForWalletId = async ({
  walletId,
}: {
  walletId: WalletId
}): Promise<PartialTransactions | ApplicationError >

// or the name you want
type PartialTransactions = {
  transactions: WalletTransaction[]
  isPartial: boolean
  // or use error 
  error?: ApplicationError

  // later we will have to add something like cursor or other related pagination fields
}
```

yes, probably we should do a log if we have an internal error (getting incoming transactions) but we will not stop the flow of the app so we can simply return Transactions (as alternative) and handle the log/notification internally

File: src/domain/ledger/index.types.d.ts
Comment (samerbuna) [inline src/domain/ledger/index.types.d.ts:14]: Extract `typeof import("./index")` into a named type here to simplify?

File: src/servers/graphql-old-server.ts
Comment (dolcalmi) [inline src/servers/graphql-old-server.ts:86]: so if we have problem with onchain incoming transactions will we give an error to the user? (we already have transactions from ledger)

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (bodymindarts) [reply src/app/wallets/get-transactions-for-wallet.ts]: I think if both a result and an error is returned its clear that it is a partial result.

File: src/app/wallets/get-transactions-for-wallet.ts
Comment (dolcalmi) [inline src/app/wallets/get-transactions-for-wallet.ts:18]: I really dont understand why we are over complicating this, just return transactions because we are throwing an error to the user if we have problems with incoming txs

Comment (bodymindarts) [reply src/app/wallets/get-transactions-for-wallet.ts:18]: Right now we are throwing an error to move forward but we want to translate and return the error properly via graphql

File: src/servers/graphql-old-server.ts
Comment (bodymindarts) [reply src/servers/graphql-old-server.ts:86]: yes we should then return both the transactions and the error via graphql... handling of error in gql layer is out of scope for this PR

**Review by vindard — APPROVED** [review]
The new `PartialResult` code was the only tricky new thing to wrap my head around, but otherwise this looks good from my last review ([relevant diff](https://github.com/GaloyMoney/galoy/compare/5644d7bf0ee93f692ba03d094ced286b59a80f06...31e76ecd0c30e2cc851d2c06eb260bb336f8e4b4))

**Review by samerbuna — APPROVED** [review]

**Pull request merged**
